src/v/crypto/digest.cc

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "base/vassert.h"
+#include "crypto/crypto.h"
+#include "internal.h"
+#include "ssl_utils.h"
+
+#include <openssl/evp.h>
+
+#include <memory>
+#include <optional>
+
+namespace crypto {
+digest_context::~digest_context() noexcept = default;
+
+class digest_context::impl {
+public:
+    explicit impl(digest_type type)
+      : _type(type)
+      , _md_ctx(internal::EVP_MD_CTX_ptr(EVP_MD_CTX_new())) {
+        if (!_md_ctx) {
+            throw internal::ossl_error("Failed to allocate new EVP_MD_CTX");
+        }
+
+        if (
+          1
+          != EVP_DigestInit_ex(
+            _md_ctx.get(), internal::get_md(type), nullptr)) {
+            throw internal::ossl_error(fmt::format(
+              "Failed to initialize digest operation for {}", type));
+        }
+    }
+
+    void update(bytes_view msg) {
+        vassert(!_is_finished, "Digest operation already finished");
```

Comment:
The reason I didn't make it r-value was because `finish` could fail if you didn't provide the appropriate sized buffer but the context would still be valid.

Reply:
Alternatively, instead of having the user provide an appropriate sized buffer, it's probably permissible to return one.  SHA512 digest (being the largest currently supported one) is only 64 bytes.